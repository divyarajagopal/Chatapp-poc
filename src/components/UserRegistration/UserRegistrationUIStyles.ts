import { StyleSheet } from 'react-native';
export const UserRegistrationStyles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  inputContainer: {
    width: '80%'
  },
  loader: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    justifyContent: 'center'
  }
});
