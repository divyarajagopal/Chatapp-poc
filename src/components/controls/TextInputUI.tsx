import * as React from 'react';
import { TextInput, StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  textInputUnderline: {
    color: 'transparent'
  },
  textInput: {
    width: '100%',
    margin: 10,
    padding: 10,
    borderColor: 'grey',
    borderWidth: 3,
    borderRadius: 5
  },
  placeholderText: {
    color: 'grey'
  },
  inputError: {
    borderColor: 'red',
    borderWidth: 3
  }
});

const TextInputUI = (props: any) => (
  <TextInput
    underlineColorAndroid={styles.textInputUnderline.color}
    placeholderTextColor={styles.placeholderText.color}
    {...props}
    style={[styles.textInput, props.error ? styles.inputError : null]}
  />
);

export default TextInputUI;
