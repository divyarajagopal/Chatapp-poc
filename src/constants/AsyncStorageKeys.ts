export enum AsyncStorageKeys {
    CHAT_SESSION = 'CHAT_SESSION'
}
