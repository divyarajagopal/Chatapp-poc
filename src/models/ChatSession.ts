export interface IChatSession {
    username?: string;
    email?: string
}