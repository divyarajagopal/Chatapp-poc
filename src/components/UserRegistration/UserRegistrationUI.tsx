import * as React from 'react';
import { Component } from 'react';
import { View, Button, ActivityIndicator, AsyncStorage } from 'react-native';
import { UserRegistrationStyles as styles } from './UserRegistrationUIStyles';
import { IUserRegistrationRender } from '../../__common__/containers/UserRegistration/UserRegistrationRender';
import { IChatSession } from '../../models/ChatSession';
import { AsyncStorageKeys } from '../../constants/AsyncStorageKeys';
import { ConnectedUserRegistration } from '../../__common__/containers/UserRegistration/UserRegistration';
import TextInputUI from '../controls/TextInputUI';
import TouchableOpacityUI from '../controls/TouchableOpacityUI';

class UserRegistrationUI extends Component<any, any> {
  constructor(props: any) {
    super(props);
  }

  public onStartChatPressed = async (renderObject: IUserRegistrationRender) => {
    renderObject.onStartChatPressed();

    const chatSessionObj: IChatSession = {
      username: renderObject.state.userName,
      email: renderObject.state.email
    };
    await AsyncStorage.setItem(
      AsyncStorageKeys.CHAT_SESSION,
      JSON.stringify(chatSessionObj)
    );
  };

  public render = () => {
    const renderLoader = (isLoaded?: boolean) => {
      let loader = <View />;
      if (!isLoaded) {
        loader = (
          <View style={styles.loader}>
            <ActivityIndicator size="large" />
          </View>
        );
      }
      return loader;
    };

    return (
      <View style={{ flex: 1 }}>
        <ConnectedUserRegistration
          navigation={this.props.navigation}
          render={(renderObj: IUserRegistrationRender) => (
            <View style={styles.container}>
              {renderLoader(renderObj.props && renderObj.props.isLoaded)}
              <View style={styles.inputContainer}>
                <TextInputUI
                  error={
                    renderObj.state.userName && !renderObj.state.isValidUserName
                  }
                  value={renderObj.state.userName}
                  placeholder="Username"
                  onChangeText={renderObj.onUsernameChangedHandler}
                />
                <TextInputUI
                  error={renderObj.state.email && !renderObj.state.isValidEmail}
                  value={renderObj.state.email}
                  placeholder="Email"
                  onChangeText={renderObj.onEmailChangedHandler}
                />
                <TouchableOpacityUI>
                  <Button
                    disabled={!renderObj.state.enableStartChatBtn}
                    title="Start Chat"
                    onPress={async () =>
                      await this.onStartChatPressed(renderObj)
                    }
                  />
                </TouchableOpacityUI>
              </View>
            </View>
          )}
        />
      </View>
    );
  };
}

export default UserRegistrationUI;
