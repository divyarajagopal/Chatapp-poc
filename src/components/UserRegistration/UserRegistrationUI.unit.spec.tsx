/**
 * @jest-environment jsdom
 */
import * as React from 'react';
import { IApplicationStoreState } from '../../__common__';
import {
  IsLoadingActionCreator,
  IsLoadedActionCreator
} from '../../__common__/redux-modules/welcome';
import { Button, AsyncStorage } from 'react-native';
import configureStore from 'redux-mock-store';
import { configure, shallow, ShallowWrapper } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import UserRegistrationUI from './UserRegistrationUI';
import { ConnectedUserRegistration } from '../../__common__/containers/UserRegistration/UserRegistration';
import { IUserRegistrationRender } from '../../__common__/containers/UserRegistration/UserRegistrationRender';
import { IChatSession } from '../../models/ChatSession';
import { AsyncStorageKeys } from '../../constants/AsyncStorageKeys';
import TextInputUI from '../controls/TextInputUI';

describe('UserRegistrationUI Component unit tests', () => {
  configure({ adapter: new Adapter() });
  const mockStore = configureStore();
  let store: any, wrapper: any, rootWrapper: any;
  beforeEach(() => {
    const applicationState: IApplicationStoreState = {
      welcomeStoreState: {
        data: [],
        errorMessage: undefined,
        isLoaded: false
      }
    };
    store = mockStore(applicationState);
    rootWrapper = shallow(<UserRegistrationUI />, {
      context: { store }
    });
    wrapper = shallow(<UserRegistrationUI />, {
      context: { store }
    });
    store.dispatch(IsLoadingActionCreator());
    store.dispatch(IsLoadedActionCreator());

    const renderParams: any = {
      state: {
        enableStartChatBtn: false,
        email: '',
        userName: '',
        isValidEmail: false,
        isValidUserName: false
      },
      props: {
        navigation: null,
        onConnecting: () => {},
        onLoaded: () => {},
        onLoading: () => {}
      },
      onEmailChangedHandler: () => {},
      onStartChatPressed: () => {},
      onUsernameChangedHandler: () => {}
    };
    wrapper = wrapper.find(ConnectedUserRegistration).renderProp('render')(
      renderParams
    );
  });

  test('should save the username and email in the Async Storage', async () => {
    const renderParams: IUserRegistrationRender = {
      state: {
        userName: 'test',
        email: 'test@test.com',
        isValidEmail: true,
        isValidUserName: true,
        enableStartChatBtn: true
      },
      onEmailChangedHandler: jest.fn(),
      onStartChatPressed: jest.fn(),
      onUsernameChangedHandler: jest.fn()
    };
    AsyncStorage.setItem = jest.fn();
    await ((rootWrapper as ShallowWrapper).instance() as UserRegistrationUI).onStartChatPressed(
      renderParams
    );
    expect(AsyncStorage.setItem).toHaveBeenCalledTimes(1);
    const chatSessionObj: IChatSession = {
      username: renderParams.state.userName,
      email: renderParams.state.email
    };
    expect(AsyncStorage.setItem).toHaveBeenCalledWith(
      AsyncStorageKeys.CHAT_SESSION,
      JSON.stringify(chatSessionObj)
    );
    expect(renderParams.onStartChatPressed).toHaveBeenCalledTimes(1);
  });

  test('should render the component as expected with store context', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have two TextInput components', () => {
    expect(wrapper.find(TextInputUI)).toHaveLength(2);
  });

  test('should have one Button component', () => {
    expect(wrapper.find(Button)).toHaveLength(1);
  });

  test('should have loaded with default Username and Email', () => {
    expect(
      wrapper.find('TextInputUI[placeholder="Username"]').get(0).props.value
    ).toBe('');
    expect(
      wrapper.find('TextInputUI[placeholder="Email"]').get(0).props.value
    ).toBe('');
  });

  test('should have loaded with disabled button with a title of `Start Chat`', () => {
    expect(
      wrapper.find('TextInputUI[placeholder="Username"]').get(0).props.value
    ).toBe('');
    const buttonProps = wrapper.find(Button).get(0).props;
    expect(buttonProps.disabled).toBeTruthy();
    expect(buttonProps.title).toBe('Start Chat');
  });
});
