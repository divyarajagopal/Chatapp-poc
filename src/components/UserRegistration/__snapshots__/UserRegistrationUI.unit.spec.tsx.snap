// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRegistrationUI Component unit tests should render the component as expected with store context 1`] = `
<Component
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Component
    style={
      Object {
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Component
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
  </Component>
  <Component
    style={
      Object {
        "width": "80%",
      }
    }
  >
    <TextInputUI
      error=""
      onChangeText={[Function]}
      placeholder="Username"
      value=""
    />
    <TextInputUI
      error=""
      onChangeText={[Function]}
      placeholder="Email"
      value=""
    />
    <TouchableOpacityUI>
      <Button
        disabled={true}
        onPress={[Function]}
        title="Start Chat"
      />
    </TouchableOpacityUI>
  </Component>
</Component>
`;
